src/audio/module_adapter/module_adapter.c

```diff
@@ -387,16 +396,15 @@ int module_adapter_prepare(struct comp_dev *dev)
 		buffer_detach(buffer, &mod->sink_buffer_list, PPL_DIR_UPSTREAM);
 		buffer_free(buffer);
 	}
-out_free:
+out_data_free:
 	for (i = 0; i < mod->num_output_buffers; i++)
 		rfree((__sparse_force void *)mod->output_buffers[i].data);
-
-	rfree(mod->output_buffers);
-
-in_free:
+in_data_free:
 	for (i = 0; i < mod->num_input_buffers; i++)
 		rfree((__sparse_force void *)mod->input_buffers[i].data);
-
+out_free:
+	rfree(mod->output_buffers);
+in_free:
 	rfree(mod->input_buffers);
 	return ret;
```

Comment:
@lyakh Ok, I've added an assignment to NULL for both `input_buffers `and `output_buffers` if their number = 0.

Reply:
Ok, yes, this fixes my concern, thanks. Now that's safe against that potential bug. Sorry, *but* - there's a "but": (1) generally if I do an if-else block with both branches present, I'd rather do
```
if (a)
	do_a();
else
	do_b();
```
than
```
if (!a)
	do_b();
else
	do_a();
```
and (2) you added code to cover that potential problem to the main code path. That means it will now always be executed - just to protect an error case. Whereas alternatively you could just place checks for the number of input and output buffers on the error path - below the `in_free` and `out_free` labels. That way your "good" path wouldn't be punished - it wouldn't need to execute any extra code, and only the error path would be "punished" - but the error path is anyway "never" executed and if it is - it's ok for it to run slightly slower. Of course - both remarks perfectly optional. My actual concern is addressed - thanks. Just some ideas to potentially improve it a bit.

Reply:
@lyakh I agree with (1).

(2): My intention was to not proceed with any memory allocation ops in case of `num_input_buffers ` or `num_output_buffers ` counts to 0. Please correct me if I'm wrong, in case of module with input or output only this part of the code will always try to allocate memory for `mod->input_buffers` or `mod->output_buffers`. Also, I don't see any checks for the number of `bytes` provided to `rzalloc()` and consecutive `malloc` operations. It looks like we will go deeper and deeper knowing we're going to allocate size = 0. I suppose penalty could be higher without those checks before allocation.

Reply:
to (2) in principle `sys_heap_aligned_alloc()` has a check for 0-byte allocation, but ok, we can keep it here too.